### ISSUE TITLE
Key-extractor called twice with StreamBridge.send

### ISSUE THREAD

**Issue by ennishol** [spring-cloud/spring-cloud-stream#2771]
Wenn a message is sent out using StreamBridge.send the custom key extractor is called twice: one time with the expected payload and the second time with the `Message<byte[]>` payload, which is unexpected.

The expected behavior is that the key extractor should be invoked only once per message.

How to reproduce:
- start the application
- curl localhost:8080/trigger
- watch the console or use the debugger on demoKeyExtractor invocation

```yaml
spring:
  cloud:
    stream:
      bindings:
        generate-out-0:
          destination: demoinput
          producer:
            partition-key-extractor-name: demoKeyExtractor
            partition-count: 16
      kafka:
        binder:
          auto-add-partitions: true
          min-partition-count: 16
```
```java
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.stream.binder.PartitionKeyExtractorStrategy;
import org.springframework.cloud.stream.function.StreamBridge;
import org.springframework.context.annotation.Bean;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@SpringBootApplication
public class CloudStreamDemoApplication {
	public static void main(String[] args) {
		SpringApplication.run(CloudStreamDemoApplication.class, args);
	}

	@Bean
	public PartitionKeyExtractorStrategy demoKeyExtractor() {
		return message -> {
			Object payload = message.getPayload();
			if (payload instanceof String p) {
				System.out.println("key extractor key is " + p);
				return p;
			}
			System.out.println("key extractor payload class is " + payload.getClass());
			return payload;
		};
	}

	@RestController
	public static class Api {
		private final StreamBridge streamBridge;

		public Api(StreamBridge streamBridge) {this.streamBridge = streamBridge;}

		@GetMapping("/trigger")
		public void name() {
			streamBridge.send("generate-out-0", "Alice");
		}
	}

}

```

**Comment by sobychacko** [comment]
@ennishol This looks like an issue we recently [addressed](https://github.com/spring-cloud/spring-cloud-stream/issues/2759). Which version are you using? Can you try it against the latest snapshots? (`4.0.4-SNAPSHOT` or `4.1.0-SNAPSHOT`). Please let us know if these versions fix this issue for you.

**Comment by ennishol** [comment]
@sobychacko  works with `2023.0.0-SNAPSHOT`

**Comment by sobychacko** [comment]
Good to hear that. Feel free to close this issue unless you have any other related concerns.

**Comment by ennishol** [comment]
fixed in `2023.0.0-SNAPSHOT` and `2022.0.4-SNAPSHOT`
